### PR TITLE
Stream buffer abort/v2

### PR DESCRIPTION
**Pull request by victorjulien**
Draft for QA/CI

### CODE REVIEW

File: src/util-streaming-buffer.c [hunk 1]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
-                to_shift->buf = ptr;
-                to_shift->buf_size = new_mem_size;
+
+            /* see if after the slide we'd overlap with the next region. If so, we need
+             * to consolidate them into one. */
+            if (to_shift->next && slide_offset + new_mem_size >= to_shift->next->stream_offset) {
+                StreamingBufferRegion *start = to_shift;
+                StreamingBufferRegion *next = start->next;
+
+                // using big as main
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:855]: next

File: src/util-streaming-buffer.c [hunk 2]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                if (start->buf_size < next->buf_size) {
+                    SCLogDebug("replace main with the next bigger region");
+
+                    const uint64_t next_re = next->stream_offset + next->buf_size;
+                    uint32_t next_mem_size = next_re - slide_offset;
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:860]: single line, next_mem_size const

File: src/util-streaming-buffer.c [hunk 3]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                    next_mem_size = ToNextMultipleOf(next_mem_size, cfg->buf_size);
+                    const uint32_t next_data_offset = next->stream_offset - slide_offset;
+                    const uint32_t prev_buf_size = next->buf_size;
+
+                    // expand "big" to include relevant part of "start"
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:865]: next

File: src/util-streaming-buffer.c [hunk 4]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                    if (GrowRegionToSize(sb, cfg, next, next_mem_size) != 0)
+                        return; // TODO what to do?
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:867]: handle error

File: src/util-streaming-buffer.c [hunk 5]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                    // post-move [XXXnextnextnext]
+                    memmove(next->buf + next_data_offset, next->buf, prev_buf_size);
+
+                    // move portion of "start" into "next"
+                    //
+                    // start: [ooooookkkkk] (o: old, k: keep)
+                    // pre-next     [xxxxxbigbigbig]
+                    // post-next    [kkkkkbigbigbig]
+                    const uint32_t start_data_offset = slide_offset - start->stream_offset;
+                    BUG_ON(start_data_offset > start->buf_size);
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:882]: debug validate + error handle

File: src/util-streaming-buffer.c [hunk 6]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+
+                    // free "next"
+                    FREE(cfg, next, sizeof(*next));
+                    sb->regions--;
+                    BUG_ON(sb->regions == 0);
+                } else {
+                    // using "main"
+
+                    // expand to include "next"
+                    const uint64_t next_re = next->stream_offset + next->buf_size;
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:903]: next_re can drop down one level as it's the same in both branches

File: src/util-streaming-buffer.c [hunk 7]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                    uint32_t start_mem_size = next_re - slide_offset;
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:904]: single line

Comment (victorjulien) [reply src/util-streaming-buffer.c:904]: start_mem_size and next_mem_size from other branch are the same, so consolidate

File: src/util-streaming-buffer.c [hunk 8]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                    start_mem_size = ToNextMultipleOf(start_mem_size, cfg->buf_size);
+
+                    if (GrowRegionToSize(sb, cfg, start, start_mem_size) != 0)
+                        return; // TODO what to do?
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:908]: error handle

File: src/util-streaming-buffer.c [hunk 9]
```diff
@@ -754,19 +845,110 @@ static inline void StreamingBufferSlideToOffsetWithRegions(
...
+                    SCLogDebug("copy next->buf %p/%u to start->buf offset %u", next->buf,
+                            next->buf_size, new_data_size);
+                    memcpy(start->buf + new_data_size, next->buf, next->buf_size);
+
+                    start->buf_size = start_mem_size;
+                    start->stream_offset = slide_offset;
+                    start->next = next->next;
+
+                    // free "next"
+
```
Comment (victorjulien) [inline src/util-streaming-buffer.c:930]: remove newline

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW1_files_sha256.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .flow.memuse | 526058088 | 615198192 | 116.94% |
| .http.memuse | 2104 | 0 | - | 

Pipeline 14295

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .flow.memuse | 519171056 | 2157879448 | 415.64% | 

Pipeline 14314

**Comment by victorjulien** [comment]
replaced by #8990

**Pull request closed without merging**